src/all/mangafire/src/eu/kanade/tachiyomi/extension/all/mangafire/MangaFire.kt

```diff
@@ -147,137 +150,115 @@ class MangaFire(
-            for ((i, it) in elements.withIndex()) {
-                it.attr("data-id", chapters[i].select("a").attr("data-id"))
-            }
-        }
-        return elements.toList()
+    private fun getAjaxRequest(ajaxType: String, mangaId: String, chapterType: String): Request {
+        return GET("$baseUrl/ajax/$ajaxType/$mangaId/$chapterType/$langCode", headers)
     }
 
     @Serializable
-    class ChapterIdsDto(
+    class AjaxReadDto(
         val html: String,
-        val title_format: String,
     )
 
-    private fun updateChapterList(manga: SManga, chapters: List<SChapter>) {
-        val request = chapterListRequest(manga.url, chapterType)
-        val response = client.newCall(request).execute()
-        val result = json.decodeFromString<ResponseDto<String>>(response.body.string()).result
-        val document = Jsoup.parse(result)
-
-        val elements = document.selectFirst(".scroll-sm")!!.children()
-        val chapterCount = chapters.size
-        if (elements.size != chapterCount) throw Exception("Chapter count doesn't match. Try updating again.")
-        val dateFormat = SimpleDateFormat("MMM dd, yyyy", Locale.US)
-        for (i in 0 until chapterCount) {
-            val chapter = chapters[i]
-            val element = elements[i]
-            val number = element.attr("data-number").toFloatOrNull() ?: -1f
-            if (chapter.chapter_number != number) throw Exception("Chapter number doesn't match. Try updating again.")
-            chapter.name = element.select(Evaluator.Tag("span"))[0].ownText()
-            val date = element.select(Evaluator.Tag("span"))[1].ownText()
-            chapter.date_upload = try {
-                dateFormat.parse(date)!!.time
-            } catch (_: Throwable) {
-                0
+    override fun chapterListParse(response: Response): List<SChapter> {
+        throw UnsupportedOperationException()
+    }
+
+    override fun fetchChapterList(manga: SManga): Observable<List<SChapter>> {
+        val path = manga.url
+        val mangaId = path.removeSuffix(VOLUME_URL_SUFFIX).substringAfterLast(".")
+        val isVolume = path.endsWith(VOLUME_URL_SUFFIX)
+
+        val type = if (isVolume) "volume" else "chapter"
+        val abbrPrefix = if (isVolume) "Vol" else "Chap"
+        val fullPrefix = if (isVolume) "Volume" else "Chapter"
+
+        val ajaxMangaList = client.newCall(getAjaxRequest("manga", mangaId, type))
+            .execute().parseAs<ResponseDto<String>>().result
+            .toBodyFragment()
+            .select(if (isVolume) ".vol-list > .item" else "li")
+
+        val ajaxReadList = client.newCall(getAjaxRequest("read", mangaId, type))
+            .execute().parseAs<ResponseDto<AjaxReadDto>>().result.html
+            .toBodyFragment()
+            .select("ul a")
+
+        val chapterList = ajaxMangaList.zip(ajaxReadList) { m, r ->
+            val link = r.selectFirst("a")!!
+            if (!r.attr("href").substringAfterLast("/").contains(type)) {
```

Comment:
Are you getting this from the fragment or path? Might be better to parse with `HttpUrl`.